### ISSUE TITLE
Add gradients to encode function

### ISSUE THREAD

**Issue by whoisjones** [UKPLab/sentence-transformers#274]
Hi, would be nice to have the possibility to get the embeddings with gradients enabled. Right now, it always runs through with torch.no_grad() in the encode function. Or is there another way to get this?

**Comment by nreimers** [comment]
You can use the forward method of SentenceTransformers. This allow to include SentenceTransformer into a larger model.

However, the forward method expects a  correctly formatted batch as input.

**Comment by whoisjones** [comment]
okay thx!

**Comment by BelhalK** [comment]
Hello,

Would love to reopen this issue as I am facing a similar one.
Basically, I am using SentenceTransformer in a larger model and indeed my input embeddings (obtained through SentenceTransformers) are detached from the computation graph and no grad can be derived.

Can this simple example where the output embeddings has no `.grad` help in solving this issue?

```
modelB = SentenceTransformer('clip-ViT-B-32-multilingual-v1')
out = modelB.encode('a sentence', convert_to_tensor=True)
print(out.grad)
>>> None
```

How can I make sure that the return value of `modelB.encode` is differentiable?

Thank you very much.

**Comment by nreimers** [comment]
@BelhalK 
This should work:
```
encoded_input = model.tokenize(your_input)
out = model(encoded_input)
```

Out will contain a dict with different features. The `sentence_embedding` feature is the final output of the model

**Comment by BelhalK** [comment]
Thank you.
Indeed they seem more differentiable.

Though, you could maybe help me understand the discrepancy in their size and output (I would expect the same embeddings in the following ways of using encode and tokenize)

```
myinput = 'a sentence'
out1 = model.encode(myinput, convert_to_tensor=True)
print(out1.shape)
print(out1[0:10])

>>> torch.Size([512])
>>> tensor([-0.0598, -0.0616, -0.1732,  0.0497, -0.3227, -0.0809, -0.2348, -1.2421,
        -0.0162,  0.2095], device='cuda:0')
```
and
```
encoded_input = model.tokenize(myinput)
out2 = model(encoded_input)
print(out2.shape)
print(out2["sentence_embedding"][0][0:10]) #printing first vector but none matches the value of above

>>> torch.Size([10, 512])
>>> tensor([ 0.0059,  0.2109,  0.1050,  0.2442, -0.0504, -0.0644, -0.2083, -0.6170,
         0.0738,  0.3470], grad_fn=<SliceBackward>)
```

The discrepancy of the values is not too concerning (it might have to do with the ordering of the tokens) but I don't understand why the output is of size 10 and not 1 for a single input.

**Comment by nreimers** [comment]
Your input has to be a list, not a single string. So add [ and ] around your input

**Comment by BelhalK** [comment]
Got it.
It worked.

Thank you a lot for the main issue!
Tokenize and predict is the way to go for backpropagation.
